apps/openmw/mwgui/sortfilteritemmodel.cpp

```diff
@@ -59,15 +59,95 @@ namespace
             if (mSortByType && left.mType != right.mType)
                 return left.mType < right.mType;
 
-            if (left.mBase.getTypeName() == right.mBase.getTypeName())
+            float result = 0;
+
+            // compare items by type
+            std::string leftName = left.mBase.getTypeName();
+            std::string rightName = right.mBase.getTypeName();
+
+            if (leftName != rightName)
+                return compareType(leftName, rightName);
+
+            // compare items by name
+            leftName = Misc::StringUtils::lowerCase(left.mBase.getClass().getName(left.mBase));
+            rightName = Misc::StringUtils::lowerCase(right.mBase.getClass().getName(right.mBase));
+
+            result = leftName.compare(rightName);
+            if (result != 0)
+                return result < 0;
+
+            // compare items by enchantment:
+            // 1. enchanted items showed before non-enchanted
+            // 2. item with lesser charge percent comes after items with more charge percent
+            // 3. item with constant effect comes before items with non-constant effects
+            int leftChargePercent = -1;
+            int rightChargePercent = -1;
+            leftName = left.mBase.getClass().getEnchantment(left.mBase);
+            rightName = right.mBase.getClass().getEnchantment(right.mBase);
+
+            if (!leftName.empty())
             {
-                std::string leftName = Misc::StringUtils::lowerCase(left.mBase.getClass().getName(left.mBase));
-                std::string rightName = Misc::StringUtils::lowerCase(right.mBase.getClass().getName(right.mBase));
+                const ESM::Enchantment* ench = MWBase::Environment::get().getWorld()->getStore().get<ESM::Enchantment>().search(leftName);
+                if (ench)
+                {
+                    if (ench->mData.mType == ESM::Enchantment::ConstantEffect)
+                        leftChargePercent = 101;
+                    else
+                        leftChargePercent = (left.mBase.getCellRef().getEnchantmentCharge() == -1) ? 100
+                            : static_cast<int>(left.mBase.getCellRef().getEnchantmentCharge() / static_cast<float>(ench->mData.mCharge) * 100);
+                }
+                else
+                    std::cerr << "Warning: Can't find enchantment '" << leftName << "' on item " << left.mBase.getCellRef().getRefId() << std::endl;
```

Comment:
I'd get rid of this warning. It'll be printed every time the inventory is sorted, multiple times even. We already have similar warnings elsewhere that should be sufficient.

Ideally, any invalid references would be checked when loading the game, then replaced with null references, so the actual game logic doesn't have to deal with invalid references, and warnings are only printed once. That's how vanilla behaves, AFAIK. We'll need quite a bit of refactoring to get there, though.